File: software/nbody.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <time.h>

#include "vga_ball.h"

int vga_ball_fd;

/* Read and print the properties color */
void print_properties_color() {
  vga_ball_arg_t vla;
  
  if (ioctl(vga_ball_fd, VGA_BALL_READ_PROPERTIES, &vla)) {
      perror("ioctl(VGA_BALL_READ_PROPERTIES) failed");
      return;
  }
  printf("%02x %02x %02x %02x\n",
	 vla.properties.x1, vla.properties.y1, vla.properties.x2, vla.properties.y2);
}

/* Set the properties color */
void set_properties_color(const vga_ball_props_t *c)
{
  vga_ball_arg_t vla;
  vla.properties = *c;
  if (ioctl(vga_ball_fd, VGA_BALL_WRITE_PROPERTIES, &vla)) {
      perror("ioctl(VGA_BALL_SET_PROPERTIES) failed");
      return;
  }
}


// Function to calculate the acceleration for each body
void getA(double** R, double G, double* M, int N, double** A) {
    double** F = (double**)malloc(N * sizeof(double*));
    for (int i = 0; i < N; i++) {
        F[i] = (double*)calloc(2, sizeof(double));
    }

    // Calculate forces for each body
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (i != j) { // Avoid calculating force with itself
                double dx = R[i][0] - R[j][0];
                double dy = R[i][1] - R[j][1];
                double distance = sqrt(dx * dx + dy * dy);
                double factor = G * M[i] * M[j] / pow(distance, 3);
                
                F[i][0] += factor * dx;
                F[i][1] += factor * dy;
            }
        }
        
        // Calculate acceleration from force: A = -F/M
        A[i][0] = -1 * (F[i][0] / M[i]);
        A[i][1] = -1 * (F[i][1] / M[i]);
    }

    // Print accelerations
    printf("Accelerations:\n");
    for (int i = 0; i < N; i++) {
        printf("Body %d: (%f, %f)\n", i+1, A[i][0], A[i][1]);
    }

    // Free allocated memory for F
    for (int i = 0; i < N; i++) {
        free(F[i]);
    }
    free(F);
}

// Main function with N-body simulation
void simulate(int N, double T, double dt, vga_ball_props_t init) {
    double t0 = 0.0;
    int timestep = 0;
    int total_timesteps = (int)(T / dt) + 1;
    
    // Gravitational constant
    double G = 10.0;
    
    // Allocate memory for positions, velocities, accelerations, and masses
    double** R = (double**)malloc(N * sizeof(double*));
    double** V = (double**)malloc(N * sizeof(double*));
    double** A = (double**)malloc(N * sizeof(double*));
    double* M = (double*)malloc(N * sizeof(double));
    
    /*
    // Allocate memory for position history (for visualization if needed)
    double*** positions = (double***)malloc(total_timesteps * sizeof(double**));
    for (int t = 0; t < total_timesteps; t++) {
        positions[t] = (double**)malloc(N * sizeof(double*));
        for (int i = 0; i < N; i++) {
            positions[t][i] = (double*)malloc(2 * sizeof(double));
        }
    }*/
    
    // Initialize arrays
    for (int i = 0; i < N; i++) {
        R[i] = (double*)calloc(2, sizeof(double));
        V[i] = (double*)calloc(2, sizeof(double));
        A[i] = (double*)calloc(2, sizeof(double));
    }
    
    // Get initial positions and masses from user
    for (int i = 0; i < N; i++) {
        printf("Body #%d - Type in X Position: ", i+1);
        scanf("%lf", &R[i][0]);
        
        printf("Body #%d - Type in Y Position: ", i+1);
        scanf("%lf", &R[i][1]);
        
        printf("Body #%d - Type in Mass: ", i+1);
        scanf("%lf", &M[i]);
    }
    
    // Calculate initial accelerations
    getA(R, G, M, N, A);

    set_properties_color(&init);

    // Main simulation loop
    while (t0 < T) {
        /*
        // Store current positions (for visualization if needed)
        for (int i = 0; i < N; i++) {
            positions[timestep][i][0] = R[i][0];
            positions[timestep][i][1] = R[i][1];
        }*/
        
        // Print current positions
        printf("Time: %.2f\n", t0);
        for (int i = 0; i < N; i++) {
            printf("Body %d position: (%.2f, %.2f)\n", i+1, R[i][0], R[i][1]);
        }
        printf("\n");
        
        // Leapfrog integrator
        
        // 1. Update velocities by half a time step
        for (int i = 0; i < N; i++) {
            V[i][0] += A[i][0] * dt / 2.0;
            V[i][1] += A[i][1] * dt / 2.0;
        }
        
        // 2. Update positions by a full time step
        for (int i = 0; i < N; i++) {
            R[i][0] += V[i][0] * dt;
            R[i][1] += V[i][1] * dt;
        }

        init.x1 = (unsigned short)R[0][0];
        init.y1 = (unsigned short)R[0][1];
        init.x2 = (unsigned short)R[1][0];
        init.y2 = (unsigned short)R[1][1];
        set_properties_color(&init);

        // 3. Recalculate accelerations
        getA(R, G, M, N, A);
        
        // 4. Update velocities by another half time step
        for (int i = 0; i < N; i++) {
            V[i][0] += A[i][0] * dt / 2.0;
            V[i][1] += A[i][1] * dt / 2.0;
        }
        
        // Update time and timestep
        t0 += dt;
        timestep++;
    }
    
    // Free all allocated memory
    for (int i = 0; i < N; i++) {
        free(R[i]);
        free(V[i]);
        free(A[i]);
    }
    free(R);
    free(V);
    free(A);
    free(M);
    
    /*
    for (int t = 0; t < total_timesteps; t++) {
        for (int i = 0; i < N; i++) {
            free(positions[t][i]);
        }
        free(positions[t]);
    }
    free(positions);
    */

}

int main() {
    vga_ball_arg_t vla;
    static const char filename[] = "/dev/vga_ball";
    vga_ball_props_t init = {320, 120, 640, 120 };

    int N = 2;   // Number of bodies
    double T = 10.0;  // Total simulation time
    double dt = 0.01; // Time step
    
    if ( (vga_ball_fd = open(filename, O_RDWR)) == -1) {
        fprintf(stderr, "could not open %s\n", filename);
        return -1;
    }

    printf("N-body simulation\n");
    printf("----------------\n");
    printf("Number of bodies: %d\n", N);
    printf("Total simulation time: %.2f\n", T);
    printf("Time step: %.2f\n\n", dt);
    
    simulate(N, T, dt, init);
    
    return 0;
}

File: software/nbody.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from matplotlib.animation import FuncAnimation


def getA(R, G, M, N):
    """Calculate acceleration using 32-bit float precision"""
    # Initialize arrays with np.float32 data type
    A = np.zeros((N, 2), dtype=np.float32)
    F = np.zeros((N, 2), dtype=np.float32)
    
    for i in range(N):
        for j in range(N):
            if i != j:
                # Direction should be FROM i TO j, so (R[j] - R[i])
                r_ij = R[j].astype(np.float32) - R[i].astype(np.float32)
                distance = np.float32(np.linalg.norm(r_ij))
                
                # Avoid division by zero or very small numbers
                if distance < np.float32(0.001):
                    distance = np.float32(0.001)
                
                # Force calculation with explicit float32 conversion
                force_scalar = np.float32(G * M[i, 0] * M[j, 0]) / np.float32(distance**3)
                force_vector = r_ij * force_scalar
                F[i] += force_vector.astype(np.float32)

        # F = ma, so a = F/m - ensure 32-bit floats
        A[i] = (F[i] / np.float32(M[i, 0])).astype(np.float32)

    return A


def main(N, T=np.float32(10.0), dt=np.float32(0.01), t0=np.float32(0.0)):
    """Main simulation function using 32-bit floats"""
    timestep = 0
    
    # Calculate number of timesteps
    num_timesteps = int(T/dt) + 1
    
    # Initialize arrays with float32 data type
    positions = np.zeros((num_timesteps, N, 2), dtype=np.float32)
    velocities = np.zeros((num_timesteps, N, 2), dtype=np.float32)
    accelerations = np.zeros((num_timesteps, N, 2), dtype=np.float32)

    G = np.float32(10.0)
    R = np.zeros((N, 2), dtype=np.float32)  # Position vectors
    V = np.zeros((N, 2), dtype=np.float32)  # Velocity vectors
    M = np.ones((N, 1), dtype=np.float32)   # Mass values

    # Randomly initialize parameters with float32
    R = np.random.uniform(-50, 50, (N, 2)).astype(np.float32)
    V = np.random.uniform(-5, 5, (N, 2)).astype(np.float32)
    M = np.random.uniform(1, 20, (N, 1)).astype(np.float32)
    
    print("Initialized bodies with random parameters (32-bit float):")
    for i in range(N):
        print(f"Body #{i+1} - Position: ({R[i,0]:.2f}, {R[i,1]:.2f}), "
              f"Velocity: ({V[i,0]:.2f}, {V[i,1]:.2f}), Mass: {M[i,0]:.2f}")
    
    # Get initial acceleration
    A = getA(R, G, M, N)

    # Setup visualization
    fig, ax = plt.subplots()
    scatter = ax.scatter(R[:,0], R[:,1], c='blue', s=M[:,0]*5, alpha=0.7)
    
    ax.set_xlim(-100, 100)
    ax.set_ylim(-100, 100)
    ax.set_title(f"N-body Simulation with {N} Bodies (32-bit float)")
    
    # Add info text
    info_text = ax.text(-95, 90, f"Time: {t0:.2f}", fontsize=10)
    
    # Main simulation loop
    while t0 < T:
        # Store current state
        positions[timestep] = R
        velocities[timestep] = V
        accelerations[timestep] = A
        
        # Update visualization
        scatter.set_offsets(R)
        info_text.set_text(f"Time: {t0:.2f}")
        plt.pause(0.01)
        
        # Velocity Verlet integration with explicit float32 conversions
        V += (A * np.float32(dt/2)).astype(np.float32)  # v(i+1/2) = v(i) + a(i)*dt/2
        R += (V * np.float32(dt)).astype(np.float32)    # x(i+1) = x(i) + v(i+1/2)*dt
        A = getA(R, G, M, N)                           # Calculate a(i+1)
        V += (A * np.float32(dt/2)).astype(np.float32)  # v(i+1) = v(i+1/2) + a(i+1)*dt/2

        t0 += dt
        timestep += 1

    plt.close()
    
    # Generate histograms after simulation completes
    create_histograms(positions, velocities, accelerations)


def create_histograms(positions, velocities, accelerations):
    """Create histograms for position, velocity and acceleration components with improved ranges"""
    # Flatten the data across all bodies and timesteps
    x_pos = positions[:,:,0].flatten()
    y_pos = positions[:,:,1].flatten()
    
    x_vel = velocities[:,:,0].flatten()
    y_vel = velocities[:,:,1].flatten()
    
    x_acc = accelerations[:,:,0].flatten()
    y_acc = accelerations[:,:,1].flatten()
    
    # Create a 2x3 grid of subplots for the histograms
    fig, axes = plt.subplots(2, 3, figsize=(15, 10))
    fig.suptitle('Distribution of Position, Velocity, and Acceleration Values (32-bit float)', fontsize=16)
    
    # Set up histogram parameters
    bins = 50
    alpha = 0.7
    
    # Function to calculate smart range based on percentiles
    def get_range(data, percentile=99):
        lower = np.percentile(data, 100-percentile)
        upper = np.percentile(data, percentile)
        # Add a bit of padding
        padding = np.float32((upper - lower) * 0.1)
        return lower - padding, upper + padding
    
    # Plot histograms with better ranges
    # X Position
    x_pos_range = get_range(x_pos)
    axes[0, 0].hist(x_pos, bins=bins, alpha=alpha, color='blue', range=x_pos_range)
    axes[0, 0].set_title('X Position')
    axes[0, 0].set_xlabel('Value')
    axes[0, 0].set_ylabel('Frequency')
    
    # X Velocity
    x_vel_range = get_range(x_vel)
    axes[0, 1].hist(x_vel, bins=bins, alpha=alpha, color='green', range=x_vel_range)
    axes[0, 1].set_title('X Velocity')
    axes[0, 1].set_xlabel('Value')
    axes[0, 1].set_ylabel('Frequency')
    
    # X Acceleration
    x_acc_range = get_range(x_acc)
    axes[0, 2].hist(x_acc, bins=bins, alpha=alpha, color='red', range=x_acc_range)
    axes[0, 2].set_title('X Acceleration')
    axes[0, 2].set_xlabel('Value')
    axes[0, 2].set_ylabel('Frequency')
    
    # Y Position
    y_pos_range = get_range(y_pos)
    axes[1, 0].hist(y_pos, bins=bins, alpha=alpha, color='blue', range=y_pos_range)
    axes[1, 0].set_title('Y Position')
    axes[1, 0].set_xlabel('Value')
    axes[1, 0].set_ylabel('Frequency')
    
    # Y Velocity
    y_vel_range = get_range(y_vel)
    axes[1, 1].hist(y_vel, bins=bins, alpha=alpha, color='green', range=y_vel_range)
    axes[1, 1].set_title('Y Velocity')
    axes[1, 1].set_xlabel('Value')
    axes[1, 1].set_ylabel('Frequency')
    
    # Y Acceleration
    y_acc_range = get_range(y_acc)
    axes[1, 2].hist(y_acc, bins=bins, alpha=alpha, color='red', range=y_acc_range)
    axes[1, 2].set_title('Y Acceleration')
    axes[1, 2].set_xlabel('Value')
    axes[1, 2].set_ylabel('Frequency')
    
    # Print ranges for each parameter
    print("\nValue Ranges (Full Data - 32-bit float):")
    print(f"X Position: [{np.min(x_pos):.2f}, {np.max(x_pos):.2f}]")
    print(f"Y Position: [{np.min(y_pos):.2f}, {np.max(y_pos):.2f}]")
    print(f"X Velocity: [{np.min(x_vel):.2f}, {np.max(x_vel):.2f}]")
    print(f"Y Velocity: [{np.min(y_vel):.2f}, {np.max(y_vel):.2f}]")
    print(f"X Acceleration: [{np.min(x_acc):.2f}, {np.max(x_acc):.2f}]")
    print(f"Y Acceleration: [{np.min(y_acc):.2f}, {np.max(y_acc):.2f}]")
    
    print("\nHistogram Display Ranges (99th percentiles):")
    print(f"X Position: [{x_pos_range[0]:.2f}, {x_pos_range[1]:.2f}]")
    print(f"Y Position: [{y_pos_range[0]:.2f}, {y_pos_range[1]:.2f}]")
    print(f"X Velocity: [{x_vel_range[0]:.2f}, {x_vel_range[1]:.2f}]")
    print(f"Y Velocity: [{y_vel_range[0]:.2f}, {y_vel_range[1]:.2f}]")
    print(f"X Acceleration: [{x_acc_range[0]:.2f}, {x_acc_range[1]:.2f}]")
    print(f"Y Acceleration: [{y_acc_range[0]:.2f}, {y_acc_range[1]:.2f}]")
    
    # Add percentage of data shown in each histogram
    for i, (ax, data, data_range) in enumerate(zip(
            [axes[0, 0], axes[0, 1], axes[0, 2], axes[1, 0], axes[1, 1], axes[1, 2]],
            [x_pos, x_vel, x_acc, y_pos, y_vel, y_acc],
            [x_pos_range, x_vel_range, x_acc_range, y_pos_range, y_vel_range, y_acc_range]
        )):
        shown_data = data[(data >= data_range[0]) & (data <= data_range[1])]
        percentage = len(shown_data) / len(data) * 100
        ax.annotate(f"Showing {percentage:.1f}% of data", xy=(0.05, 0.95), xycoords='axes fraction',
                    fontsize=8, ha='left', va='top', bbox=dict(boxstyle='round', fc='white', alpha=0.8))
    
    plt.tight_layout()
    plt.show()

    # Add a new visualization to show float32 precision effects
    plot_precision_effects(positions, velocities, accelerations)

def plot_precision_effects(positions, velocities, accelerations):
    """Create plots to visualize the effects of float32 precision"""
    # Take the last 20% of data for analysis
    start_idx = int(positions.shape[0] * 0.8)
    
    # Get position differences between timesteps
    pos_diffs = positions[1:] - positions[:-1]
    pos_diffs_magnitude = np.sqrt(pos_diffs[:,:,0]**2 + pos_diffs[:,:,1]**2)
    
    # Get velocity differences between timesteps
    vel_diffs = velocities[1:] - velocities[:-1]
    vel_diffs_magnitude = np.sqrt(vel_diffs[:,:,0]**2 + vel_diffs[:,:,1]**2)
    
    # Create figure for precision analysis
    fig, axes = plt.subplots(2, 1, figsize=(12, 10))
    fig.suptitle('32-bit Float Precision Effects', fontsize=16)
    
    # Plot position differences
    for i in range(pos_diffs_magnitude.shape[1]):  # For each body
        axes[0].plot(pos_diffs_magnitude[-100:, i], label=f"Body {i+1}")
    
    axes[0].set_title('Position Change Between Timesteps (Last 100)')
    axes[0].set_xlabel('Timestep')
    axes[0].set_ylabel('Position Change Magnitude')
    axes[0].legend()
    axes[0].grid(True)
    
    # Plot velocity differences
    for i in range(vel_diffs_magnitude.shape[1]):  # For each body
        axes[1].plot(vel_diffs_magnitude[-100:, i], label=f"Body {i+1}")
    
    axes[1].set_title('Velocity Change Between Timesteps (Last 100)')
    axes[1].set_xlabel('Timestep')
    axes[1].set_ylabel('Velocity Change Magnitude')
    axes[1].legend()
    axes[1].grid(True)
    
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    # Get only the number of bodies as input
    try:
        n_bodies = int(input("Enter number of bodies to simulate: "))
        if n_bodies < 1:
            print("Number of bodies must be at least 1. Using default of 5.")
            n_bodies = 5
    except ValueError:
        print("Invalid input. Using default of 5 bodies.")
        n_bodies = 5
        
    main(n_bodies)
